ipa-core/src/protocol/hybrid/agg.rs

```diff
@@ -0,0 +1,232 @@
+use std::collections::HashMap;
+
+use futures::{stream, FutureExt, StreamExt, TryStreamExt};
+
+use crate::{
+    error::Error,
+    ff::{boolean::Boolean, boolean_array::BooleanArray, ArrayAccess},
+    helpers::TotalRecords,
+    protocol::{
+        boolean::step::EightBitStep,
+        context::{
+            dzkp_validator::{DZKPValidator, TARGET_PROOF_SIZE},
+            Context, DZKPContext, DZKPUpgraded, MaliciousProtocolSteps, ShardedContext,
+            UpgradableContext,
+        },
+        hybrid::step::{AggregateReportsStep, HybridStep},
+        ipa_prf::boolean_ops::addition_sequential::integer_add,
+        BooleanProtocols,
+    },
+    report::hybrid::{AggregateableHybridReport, PrfHybridReport},
+    secret_sharing::{replicated::semi_honest::AdditiveShare as Replicated, SharedValue},
+    seq_join::{assert_send, seq_join, SeqJoin},
+};
+
+/// This protocol is used to aggregate `PRFHybridReports` and returns `AggregateableHybridReports`.
+/// It groups all the reports by the PRF of the `match_key`, finds all reports from `match_keys`
+/// with that provided exactly 2 reports, then adds those 2 reports.
+pub async fn aggregate_reports<BK, V, C>(
+    ctx: C,
+    reports: Vec<PrfHybridReport<BK, V>>,
+) -> Result<Vec<AggregateableHybridReport<BK, V>>, Error>
+where
+    C: UpgradableContext + ShardedContext,
+    BK: SharedValue + BooleanArray,
+    V: SharedValue + BooleanArray,
+    Replicated<Boolean>: BooleanProtocols<DZKPUpgraded<C>>,
+{
+    let mut reports_by_matchkey = HashMap::with_capacity(reports.len() / 2);
+
+    // build a hashmap of match_key -> ([AggregateableHybridReport;2], count)
+    // if count ever exceeds 2, we drop reports, but keep counting
+    // an honest client and report collector will only submit
+    // one report with a breakdown key and one report with a value.
+    // if there are less, it's unattributed. if more, something went wrong.
+    for report in reports {
+        let match_key = report.match_key;
+        let entry = reports_by_matchkey.entry(match_key).or_insert((
+            [
+                AggregateableHybridReport::<BK, V>::ZERO,
+                AggregateableHybridReport::<BK, V>::ZERO,
+            ],
+            0,
+        ));
+        if entry.1 == 0 {
+            // If the count is 0, replace the first element with the new report
+            entry.0[0] = report.into();
+            entry.1 += 1;
+        } else if entry.1 == 1 {
+            // If the count is 1, replace the second element with the new report
+            entry.0[1] = report.into();
+            entry.1 += 1;
+        } else {
+            // If the count is 2 or more, increment the counter and drop the report
+            entry.1 += 1;
```

Comment:
You could define something like this as a helper to simplify this function:

```
enum MatchEntry {
    Single(Report),
    Pair(Report, Report),
    MoreThanTwo,
}
```

Whether or not you use the enum, you can match on `reports_by_matchkey.entry(match_key)` to avoid creating a zero entry temporarily. If `Vacant`, then insert an entry with one report, otherwise (`Occupied`), add a second report or mark as overflowed.